src/show.jl

```diff
@@ -134,77 +148,84 @@ function Base.show(io::IO, ::MIME"text/plain", obj::Union{File,Group,Dataset,Att
     end
 end
 
-function _tree_icon(obj)
-    if SHOW_TREE_ICONS[]
-        return obj isa Attribute ? "🏷️" :
-               obj isa Group ? "📂" :
-               obj isa Dataset ? "🔢" :
-               obj isa Datatype ? "📄" :
-               obj isa File ? "🗂️" :
-               "❓"
-    else
-        return obj isa Attribute ? "[A]" :
-               obj isa Group ? "[G]" :
-               obj isa Dataset ? "[D]" :
-               obj isa Datatype ? "[T]" :
-               obj isa File ? "[F]" :
-               "[?]"
-    end
-end
+_tree_icon(::Type{Attribute}) = SHOW_TREE_ICONS[] ? "🏷️" : "[A]"
+_tree_icon(::Type{Group})     = SHOW_TREE_ICONS[] ? "📂" : "[G]"
+_tree_icon(::Type{Dataset})   = SHOW_TREE_ICONS[] ? "🔢" : "[D]"
+_tree_icon(::Type{Datatype})  = SHOW_TREE_ICONS[] ? "📄" : "[T]"
+_tree_icon(::Type{File})      = SHOW_TREE_ICONS[] ? "🗂️" : "[F]"
+_tree_icon(::Type)            = SHOW_TREE_ICONS[] ? "❓" : "[?]"
+_tree_icon(obj) = _tree_icon(typeof(obj))
 _tree_icon(obj::Attributes) = _tree_icon(obj.parent)
 
 _tree_head(io::IO, obj) = print(io, _tree_icon(obj), " ", obj)
 _tree_head(io::IO, obj::Datatype) = print(io, _tree_icon(obj), " HDF5.Datatype: ", name(obj))
 
-function _tree_children(parent::Union{File, Group}, attributes::Bool)
-    names = keys(parent)
-    objs  = Union{Object, Attribute}[parent[n] for n in names]
-    if attributes
-        attrn = keys(HDF5.attributes(parent))
-        attro = Union{Object, Attribute}[HDF5.attributes(parent)[n] for n in attrn]
-        names = append!(attrn, names)
-        objs  = append!(attro, objs)
+_tree_count(parent::Union{File,Group}, attributes::Bool) =
+    length(parent) + (attributes ? length(HDF5.attributes(parent)) : 0)
+_tree_count(parent::Dataset, attributes::Bool) =
+    attributes ? length(HDF5.attributes(parent)) : 0
+_tree_count(parent::Attributes, _::Bool) = length(parent)
+_tree_count(parent::Union{Attribute,Datatype}, _::Bool) = 0
+
+function _show_tree(io::IO, obj::Union{File,Group,Dataset,Datatype,Attributes,Attribute}, indent::String="";
+                    attributes::Bool = true, depth::Int = 1)
+    isempty(indent) && _tree_head(io, obj)
+    isvalid(obj) || return
+
+    INDENT = "   "
+    PIPE   = "│  "
+    TEE    = "├─ "
+    ELBOW  = "└─ "
+
+    counter = 0
+    nchildren = _tree_count(obj, attributes)
+
+    @inline function childstr(io, n, more=" ")
+        print(io, "\n", indent, ELBOW * "(", n, more, n == 1 ? "child" : "children", ")")
     end
-    return (names, objs)
-end
-function _tree_children(parent::Dataset, attributes::Bool)
-    names = String[]
-    objs = Union{Object, Attribute}[parent[n] for n in names]
+    @inline function depth_check()
+        counter += 1
+        if counter > max(2, SHOW_TREE_MAX_CHILDREN[] ÷ depth)
```

Comment:
Is your intention that this print  at depth 1 this prints 50, then 25, 12, 6, 3, 2   children at their respective depths?

Reply:
Yeah, something like that. I thought that it would balance wanting to see a reasonable number of direct children on the printed object while still avoiding very long outputs if your 50 direct children each have 50 children of their own (repeated down to the max-depth levels).